File: marketdna-web/src/types/markov_options.ts
export const REGIMES = [
  'Strong Uptrend',
  'Volatile Bull',
  'Sideways Quiet',
  'Sideways Volatile',
  'Steady Bear',
  'Volatile Bear',
] as const

export type Regime = typeof REGIMES[number]

export const REGIME_COLOR: Record<Regime, string> = {
  'Strong Uptrend':   '#22c55e',
  'Volatile Bull':    '#86efac',
  'Sideways Quiet':   '#94a3b8',
  'Sideways Volatile':'#f59e0b',
  'Steady Bear':      '#ef4444',
  'Volatile Bear':    '#7f1d1d',
}

export const REGIME_BG: Record<Regime, string> = {
  'Strong Uptrend':   'rgba(34,197,94,0.12)',
  'Volatile Bull':    'rgba(134,239,172,0.10)',
  'Sideways Quiet':   'rgba(148,163,184,0.10)',
  'Sideways Volatile':'rgba(245,158,11,0.12)',
  'Steady Bear':      'rgba(239,68,68,0.12)',
  'Volatile Bear':    'rgba(127,29,29,0.20)',
}

export interface RegimeMonth {
  month: string
  regime: Regime
  adx: number
  rsi: number
  pct_vs_sma50: number
  monthly_ret: number
  hv_ratio: number
}

export interface TransitionMatrix {
  regimes: string[]
  matrix: number[][]
  counts: number[][]
}

export interface OptionsStrategy {
  primary: string
  alternatives: string[]
  avoid: string[]
  rationale: string
  iv_action: 'Buy' | 'Sell'
}

export interface RegimeForecast {
  current_regime: Regime
  probabilities: Record<Regime, number>
  dominant_regime: Regime
  dominant_probability: number
  tail_risk_regime: Regime | null
  tail_risk_probability: number | null
  strategy: OptionsStrategy
}

export interface MarkovOptionsResult {
  symbol: string
  history: RegimeMonth[]
  matrix: TransitionMatrix
  forecast: RegimeForecast
  n_months: number
  computed_at: string
}

export interface MarketRegimeItem {
  symbol: string
  current_regime: Regime
  dominant_next: Regime
  dominant_prob: number
  primary_strategy: string
  iv_action: 'Buy' | 'Sell'
}

export interface MarketMarkovResult {
  items: MarketRegimeItem[]
  regime_distribution: Record<Regime, number>
  dominant_market_regime: Regime
  scanned_at: string
}

File: marketdna-backend/app/models/markov_options.py
"""Pydantic models for Markov Regime + Options Strategy Intelligence."""
from pydantic import BaseModel


REGIMES: list[str] = [
    "Strong Uptrend",
    "Volatile Bull",
    "Sideways Quiet",
    "Sideways Volatile",
    "Steady Bear",
    "Volatile Bear",
]


class RegimeMonth(BaseModel):
    month: str           # "2024-05"
    regime: str
    adx: float
    rsi: float
    pct_vs_sma50: float
    monthly_ret: float
    hv_ratio: float      # HV20 / median_HV20 over history


class TransitionMatrix(BaseModel):
    regimes: list[str]
    matrix: list[list[float]]  # 6×6 blended probabilities
    counts: list[list[int]]    # 6×6 raw observed counts


class OptionsStrategy(BaseModel):
    primary: str
    alternatives: list[str]
    avoid: list[str]
    rationale: str
    iv_action: str       # "Buy" or "Sell"


class RegimeForecast(BaseModel):
    current_regime: str
    probabilities: dict[str, float]  # regime → next-month probability
    dominant_regime: str
    dominant_probability: float
    tail_risk_regime: str | None      # second most likely if ≥15%
    tail_risk_probability: float | None
    strategy: OptionsStrategy


class MarkovOptionsResult(BaseModel):
    symbol: str
    history: list[RegimeMonth]
    matrix: TransitionMatrix
    forecast: RegimeForecast
    n_months: int
    computed_at: str


class MarketRegimeItem(BaseModel):
    symbol: str
    current_regime: str
    dominant_next: str
    dominant_prob: float
    primary_strategy: str
    iv_action: str


class MarketMarkovResult(BaseModel):
    items: list[MarketRegimeItem]
    regime_distribution: dict[str, int]  # current regime counts
    dominant_market_regime: str
    scanned_at: str
